File: 79-word-search/word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int m=board.length,n=board[0].length;
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(word.charAt(0)==board[i][j])
                {
                    if(dfs(0,i,j,word,board,m,n))
                    return true;
                }
            }
        }
        return false;
    }
    int di[]={-1,0,1,0};
    int dj[]={0,1,0,-1};
    public boolean dfs(int idx,int i,int j,String w,char b[][],int m,int n)
    {
        if(w.length()==idx)
        return true;
        if(i<0 || i>=m || j<0 || j>=n || b[i][j]!=w.charAt(idx))
        return false;
        char temp=b[i][j];
        b[i][j]='#';
        for(int k=0;k<4;k++)
        {
            if(dfs(idx+1,i+di[k],j+dj[k],w,b,m,n))
            return true;
        }
        b[i][j]=temp;
        return false;
    }
}

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m,n=len(board),len(board[0])
        di=[-1,0,1,0]
        dj=[0,1,0,-1]
        ans=True
        def dfs(idx,i,j):
            if idx==len(word):
                return True
            if i<0 or i>=m or j<0 or j>=n or board[i][j]!=word[idx]:
                return False
            temp,board[i][j]=board[i][j],'#'
            for k in range(4):
                ni,nj=i+di[k],j+dj[k]
                if dfs(idx+1,ni,nj):
                    return True
            board[i][j]=temp
            return False
        for i in range(m):
            for j in range(n):
                if word[0]==board[i][j]:
                    if dfs(0,i,j):
                        return True
        return False